decoder/src/frame.rs

```diff
@@ -70,68 +68,122 @@ impl<'t> Frame<'t> {
     pub fn index(&self) -> u64 {
         self.index
     }
-}
-
-pub struct DisplayMessage<'t> {
-    format: &'t str,
-    args: &'t [Arg<'t>],
-}
 
-impl fmt::Display for DisplayMessage<'_> {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        let args = format_args(self.format, self.args, None);
-        f.write_str(&args)
+    fn format_args(&self, format: &str, args: &[Arg], parent_hint: Option<&DisplayHint>) -> String {
+        self.format_args_real(format, args, parent_hint).unwrap() // cannot fail, we only write to a `String`
     }
-}
 
-/// Prints a `Frame` when formatted via `fmt::Display`, including all included metadata (level,
-/// timestamp, ...).
-pub struct DisplayFrame<'t> {
-    frame: &'t Frame<'t>,
-    colored: bool,
-}
-
-impl fmt::Display for DisplayFrame<'_> {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        let level = if self.colored {
-            match self.frame.level {
-                Level::Trace => "TRACE".dimmed().to_string(),
-                Level::Debug => "DEBUG".normal().to_string(),
-                Level::Info => "INFO".green().to_string(),
-                Level::Warn => "WARN".yellow().to_string(),
-                Level::Error => "ERROR".red().to_string(),
-            }
-        } else {
-            match self.frame.level {
-                Level::Trace => "TRACE".to_string(),
-                Level::Debug => "DEBUG".to_string(),
-                Level::Info => "INFO".to_string(),
-                Level::Warn => "WARN".to_string(),
-                Level::Error => "ERROR".to_string(),
-            }
-        };
+    fn format_args_real(
```

Comment:
(lots of churn in this file; I'm going to just trust you on these changes)

Reply:
Most of them are mechanical to thread `&Table` through